### ISSUE TITLE
Command substitution in pipe to less may stop job

### ISSUE THREAD

**Issue by krobelus** [fish-shell/fish-shell#6300]
Sometimes (about every other attempt) the following command is immediately stopped (and sent to the background).

```fish
begin echo (cat /dev/null); end | less
```

Seems to be race condition because two processes are writing to the terminal. The job stops because of SIGTTOU. It never happens without the `begin ... end`or when adding a sleep:

```fish
begin echo (sleep .1; cat /dev/null); end | less
```

Related to at least #1362.

**Comment by krobelus** [comment]
Introduced in 3bbee06248b073fc30ae5cc8e418c943c8cd6e11 which avoids having to buffer the output of the left process, see #1396.
Looks like first `less` is executed, and then `cat` which tries to write to the terminal (?) and receives `SIGTTOU`.

**Comment by ridiculousfish** [comment]
Nice find. This stuff (terminal ownership) is very delicate, if you have a fix please make a PR.

I suspect what's happening here is:

1. `cat` is its own pgroup leader. fish grants it ownership of the terminal (`tcsetpgrp`).
2. `less`  lives in fish's pgroup and claims the terminal via `tcsetpgrp`.

I would expect that `less` may be getting SIGTTOU on the tcsetpgrp call, not on a write. I'm not sure why `cat` would get it.

When real concurrent execution lands the "deferred process" hack will go away so a complicated fix is probably not worth it at this point.

**Comment by krobelus** [comment]
Fixed by 938b6838951e24dff631a9dce8c5131ef5940f4b
